base/compiler/typeinfer.jl

```diff
@@ -1007,48 +1020,110 @@ function typeinf_frame(interp::AbstractInterpreter, mi::MethodInstance, run_opti
     return frame
 end
 
+# N.B.: These need to be aligned with the C side headers
+"""
+    SOURCE_MODE_NOT_REQUIRED
+
+Indicates to inference that the source is not required and the only fields
+of the resulting `CodeInstance` that the caller is interested in are types
+and effects. Inference is still free to create a CodeInstance with source,
+but is not required to do so.
+"""
+const SOURCE_MODE_NOT_REQUIRED = 0x0
+
+"""
+    SOURCE_MODE_ABI
+
+Indicates to inference that it should return a CodeInstance that can
+either be `->invoke`'d (because it has already been compiled or because
+it has constabi) or one that can be made so by compiling its `->inferred`
+field.
+
+N.B.: The `->inferred` field is volatile and the compiler may delete it
+before setting the `->invoke` field. To ensure that the resulting CodeInstance
+maintains the required invariants, the codegen lock needs to be held.
```

Comment:
That lock doesn't maintain invariants for code outside of the lock, as well as the problem of being slated for complete deletion soon, so this cannot be required to be held.

Reply:
Ok, I'll remove the bit about the lock and leave the warning that the source may go away.